File: services/experimentation-portal/src/api/experiments/model/experiment.ts
import type { Metric } from "../../metricsCatalog";

export type CreateExperimentRequestBody = {
  name: string;
  feature_flag_id: string;
  start_time: Date;
  end_time: Date;
  hypothesis: string;
  description: string;
  variants: Array<CreateExperimentVariant>;
  metrics: Array<CreateExperimentMetric>;
};

export type CreateExperimentVariant = {
  name: string;
  key: string;
  upper_bound: number;
  lower_bound: number;
  type: VariantType;
};

export type CreateExperimentMetric = {
  metric_id: string;
  type: CreateExperimentMetricRole;
  direction: CreateExperimentMetricDirection;
  mde?: number;
  nim?: number;
};

export type CreateExperimentMetricRole =
  | "success"
  | "guardrail"
  | "deterioration"
  | "quality";
export type VariantType = "control" | "treatment";
export type CreateExperimentMetricDirection =
  | "increase"
  | "decrease"
  | "neutral";

export type ExperimentStatus =
  | "aa-planned"
  | "aa"
  | "aa-complete"
  | "ab-planned"
  | "ab"
  | "ab-complete";

export type ExperimentResponse = {
  id: string;
  name: string;
  status: ExperimentStatus;
  created_at: string;
  feature_flag_id: string;
  start_time?: Date;
  end_time?: Date;
  aa_start_time: Date;
  aa_end_time: Date;
  unique_salt: string;
  hypothesis: string;
  description: string;
  total_required_sample_size?: number;
  metrics: Array<ExperimentMetricResponse>;
  variants: Array<ExperimentVariantResponse>;
};

export type ExperimentMetricResponse = {
  metric_details: Metric;
  role: CreateExperimentMetricRole;
  direction: CreateExperimentMetricDirection;
  mde?: number;
  nim?: number;
};

export type ExperimentVariantResponse = {
  variant_key: string;
  upper_bound: number;
  lower_bound: number;
  variantType: VariantType;
};
export type RequiredSampleSizeResponse = {
  total_required_sample_size: number;
  sample_size_per_variant: Record<string, number>;
};

export type UpdateExperimentPhaseRequest = {
  start_time?: Date;
  end_time?: Date;
  bucket_allocation?: number;
};

File: libs/prismdbmodels/model/experiment/experiment.go
package experiment

import (
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5/pgtype"
)

type Experiment struct {
	ID                      uuid.UUID           `json:"id"`
	Name                    string              `json:"name"`
	Status                  ExperimentStatus    `json:"status"`
	CreatedAt               pgtype.Timestamp    `json:"created_at"`
	FeatureFlagID           string              `json:"feature_flag_id"`
	StartTime               pgtype.Timestamp    `json:"start_time"`
	EndTime                 pgtype.Timestamp    `json:"end_time"`
	AAStartTime             time.Time           `json:"aa_start_time"`
	AAEndTime               time.Time           `json:"aa_end_time"`
	Cancelled               bool                `json:"cancelled"`
	UniqueSalt              string              `json:"unique_salt"`
	Hypothesis              string              `json:"hypothesis"`
	Description             string              `json:"description"`
	TotalRequiredSampleSize *int                `json:"total_required_sample_size"`
	Metrics                 []ExperimentMetric  `json:"experiment_metrics"`
	Variants                []ExperimentVariant `json:"experiment_variants"`
}
